### PR TITLE
feat(nix-lib): initial support for import-from-derivation

### LINKED ISSUE #20: generate expression on the fly
Hi,

I am currently using moretea/yarn2nix, and was trying to figure out how I would use this.
Running yarn2nix on a lockfile does not need a network connection, right? In this case, would it be possible to build the Nix expression from the lockfile on the fly?

I am thinking about a Nix function which takes the yarn.lock file as an input and returns a derivation with the resulting node_modules.

It looks like something similar might already exist in nix-lib. Are there examples of how to use the nix-lib?

### PR DESCRIPTION
**Pull request by sternenseemann**
This pull request contains the original base change of #27 plus a simple test on top of it sanity-checking the README example.

In that test we:

* Parse a `yarn.lock` file containing some common dependencies and generate an `npm-deps.nix` file
* Parse a `package.json` file and generate an `template.nix` file (or `npm-package.nix`).
* Use those two and `buildNodeDeps` as well as `callTemplate` and `buildNodePackage` to build the resulting package.

### CODE REVIEW

File: nix-lib/default.nix [hunk 1]
```diff
@@ -109,6 +109,40 @@ let
...
 
+  # Build nix expression of dependencies based on given `yarnLock`
+  # and directly `callPackage` it. Using this can avoid the need to
+  # check in a generated nix expression. The resulting attrSet can
+  # be used as input for `buildNodeDeps`:
+  #
+  # ```
+  # buildNodeDeps (buildCallDeps { yarnLock = ./yarn.lock; })
+  # ```
+  buildCallDeps = { name ? "npm-deps.nix", yarnLock }:
```
Comment (sternenseemann) [inline nix-lib/default.nix]: Probably should name `buildCallDeps` something different

**Comment by sternenseemann** [comment]
Rebased against `master` and now using `yarn2nix --offline` to build the dependency nix expression.

File: nix-lib/default.nix (hunk 1 above)
Comment (Profpatsch) [reply nix-lib/default.nix]: Yeah, the name is not great.

`haskellPackages` uses `callCabal2nix`, so let’s use `callYarn2Nix`?

Comment (sternenseemann) [reply nix-lib/default.nix]: Maybe a bit unclear in distinction to the `--template` version? `callYarnLock2Nix` too cumbersome on the other hand.

**Comment by Profpatsch** [comment]
Should we also add `callYarnLockNix` and `callPackageJsonNix` here, which are the non-IfD versions of the functions?

**Comment by Profpatsch** [comment]
Also let’s pull `callPackage` into these functions I guess, so that you get everything with one call.

**Comment by Profpatsch** [comment]
Dumping my thoughts about the restructuring of the API (sorry German):

```
01:06:59 Profpatsch | Also das callPackage können wir glaub eh reinziehen.
01:07:18 Profpatsch | callYarnLock
01:07:26 Profpatsch | callYarnLockNix
01:07:35 Profpatsch | callPackageJson
01:07:40 Profpatsch | callPackageJsonNix
01:07:56 Profpatsch | dann ist die ifd-Funktion (mit besserer usability) die kürzere
01:08:35 Profpatsch | Und buildYarnPackage als toplevel-Hilfsfunktion
01:08:51 Profpatsch | Dann ist das schonmal vorbereitet für buildNpmPackage
01:08:57 Profpatsch | Wenn wir npm lockfiles einbauen
01:09:11 Profpatsch | Und buildYarnPackageNix
01:09:25 Profpatsch | Macht das Sinn?
01:10:02 Profpatsch | Dann gibt es später auch callNpmLock und callNpmLockNix
01:10:16 Profpatsch | Und alle generieren das gleiche Format
01:10:32 Profpatsch | Der ultimative Javascript builder quasi
```

**Comment by sternenseemann** [comment]
> Should we also add `callYarnLockNix` and `callPackageJsonNix` here, which are the non-IfD versions of the functions?

Wouldn't they just be aliases to `callPackage`?

File: nix-lib/default.nix [hunk 2]
```diff
@@ -109,6 +109,46 @@ let
...
+  # buildNodeDeps (callYarnLock ./yarn.lock {})
+  # ```
+  callYarnLock = yarnLock: { name ? "yarn.lock.nix" }:
+    pkgs.callPackage (pkgs.runCommand name {
+      # faster to build locally, see also note at linkNodeDeps
+      allowSubstitutes = false;
+      preferLocalBuild = true;
+    } ''
+      ${yarn2nix}/bin/yarn2nix --offline ${yarnLock} > $out
+    '') { };
```
Comment (Profpatsch) [inline nix-lib/default.nix:129]: I think the last {} has to be the override attrset, no? Aka we shouldn’t just make it a `{}` but pass it through.

```
callYarnLock {
  name = "yarn.lock.nix";
  lock = ./yarn.lock;
} {}
```

Comment (Profpatsch) [reply nix-lib/default.nix:129]: Nvm, the attrset is just `{ fetchurl, fetchgit }`.

**Review by Profpatsch — APPROVED** [review]

**Pull request merged**